packages/snap-client/src/Client/apis/Abstract.test.ts

```diff
@@ -0,0 +1,265 @@
+			origin: 'https://searchspring.com',
+			fetchApi: global.window.fetch,
+			headers: customHeaders,
+			maxRetry: 2,
+			cache: CustomCacheConfig,
+		};
+		let api2 = new API(new ApiConfiguration(config2));
+
+		const contextWithQuery = { ...context, query: { key: 'value' } };
+		// @ts-ignore
+		const params = api2.createFetchParams(contextWithQuery);
+		expect(params.url).toBe('https://searchspring.com/api/v1/autocomplete?key=value');
+	});
+
+	it('can use cacheKey', async () => {
+		let config: ApiConfigurationParameters = {
+			origin: 'https://searchspring.com',
+			fetchApi: global.window.fetch,
+			headers: customHeaders,
+		};
+
+		let api;
+		expect(() => {
+			api = new API(new ApiConfiguration(config));
+		}).not.toThrow();
+
+		const body = { siteId: '8uyt2m' };
+		const context = {
+			path: '/api/v1/autocomplete',
+			method: 'POST',
+			headers: config.headers,
+			body: body,
+		};
+
+		//can use the cacheKey
+		const cacheKey = '/api/v1/autocomplete' + JSON.stringify(body);
+
+		let request;
+		request = await api.request(context, cacheKey);
+
+		expect(fetchfn).toHaveBeenCalled();
+
+		fetchfn.mockClear();
+
+		//handles 200s
+		const fetchfn200 = jest
+			.spyOn(global.window, 'fetch')
+			.mockImplementation(() => Promise.resolve({ status: 200, json: () => Promise.resolve({ found: true }) } as unknown as Response));
+
+		const setCache = jest.fn();
+		const getCache = jest.fn(() => Promise.resolve({ foundInCache: true }));
+
+		api.cache.set = setCache;
+		api.cache.get = getCache;
+
+		//no cache key
+		request = await api.request(context);
+		expect(fetchfn200).toHaveBeenCalled();
+		expect(request).toEqual({ found: true });
+		expect(setCache).not.toHaveBeenCalled();
+		expect(getCache).not.toHaveBeenCalled();
+
+		fetchfn200.mockClear();
+
+		//with a cache key
+		request = await api.request(context, cacheKey);
+
+		expect(request).not.toEqual({ found: true });
+		expect(request).toEqual({ foundInCache: true });
+
+		expect(setCache).not.toHaveBeenCalled();
+		expect(getCache).toHaveBeenCalled();
+
+		fetchfn200.mockClear();
+	});
+
+	it('can handle 429s', async () => {
+		let config: ApiConfigurationParameters = {
+			origin: 'https://searchspring.com',
+			fetchApi: global.window.fetch,
+			headers: customHeaders,
+			maxRetry: 2,
+		};
+
+		let api;
+		expect(() => {
+			api = new API(new ApiConfiguration(config));
+		}).not.toThrow();
+
+		const body = { siteId: '8uyt2m' };
+		const context = {
+			path: '/api/v1/autocomplete',
+			method: 'POST',
+			headers: config.headers,
+			body: body,
+		};
+
+		//can handle 429s and retry correct amount of times.
+		const fetchfn429 = jest
+			.spyOn(global.window, 'fetch')
+			.mockImplementation(() => Promise.resolve({ status: 429, json: () => Promise.resolve({ broken: true }) } as unknown as Response));
+
+		await expect(async () => {
+			await api.request(context);
+		}).rejects.toBe(429);
+
+		expect(fetchfn429).toHaveBeenCalledTimes(config.maxRetry + 1);
```

Comment:
Perhaps the reason for this being +1 has to do with the retryCount starting at 1 instead of 0 (as mentioned earlier)?

Reply:
this +1 logic makes sense to me here. You set the max retry count to 3, expect it to make 4 calls. The config value is maxRetry, not maxAttempts. 3 retries + your initial attempt.